### PR TITLE
feat: add a dht stat command

### PR DESCRIPTION
**Pull request by Stebalien**
Currently, it just prints out the routing tables.

TODO

* [ ] ~Tests~ It's a debugging command.
* [x] Maybe better output formatting?

### CODE REVIEW

File: core/commands/dht.go [hunk 1]
```diff
@@ -39,13 +45,191 @@ var DhtCmd = &cmds.Command{
...
+	Name    string
+	Buckets []dhtBucket
+}
+
+type dhtBucket struct {
+	LastRefresh string
+	Peers       []dhtPeerInfo
+}
+
+var statDhtCmd = &cmds.Command{
```
Comment (willscott) [inline core/commands/dht.go:74]: given the level of information here, is it more sensible to have this live as `ipfs diag dht`?

File: core/commands/dht.go [hunk 2]
```diff
@@ -39,13 +45,191 @@ var DhtCmd = &cmds.Command{
...
+	Helptext: cmds.HelpText{
+		Tagline: "Returns stats about the DHTs",
```
Comment (willscott) [inline core/commands/dht.go:76]:
```suggestion
		Tagline: "Returns statistics about the DHT",
```

File: core/commands/dht.go [hunk 3]
```diff
@@ -39,13 +45,191 @@ var DhtCmd = &cmds.Command{
...
+
+		return nil
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, out dhtStat) error {
+			tw := tabwriter.NewWriter(w, 4, 4, 2, ' ', 0)
+			defer tw.Flush()
+
+			// Regex to remove fraction from duration.
+			re := regexp.MustCompile(`\.\d+`)
```
Comment (willscott) [inline core/commands/dht.go:169]: declare as variable outside of function?

File: core/commands/dht.go (hunk 1 above)
Comment (Stebalien) [reply core/commands/dht.go:74]: Possibly? I'm trying to mirror `ipfs bitswap stat` but I agree that lumping this all under the `diag` command probably makes more sense (and really, that's where the `ipfs bitswap` commands should live as well.

Comment (Stebalien) [reply core/commands/dht.go:74]: Actually, I'm not sure. `diag` is mostly for true diagnostics (running commands, system information, possibly pprof dumps in the future, etc.).

**Comment by willscott** [comment]
It seems fine to have this as a DHT sub-command. I think the only question is if it's worth having a warning or otherwise making it clear that the format and underlying implementation-specific info that's showing up may change and this isn't a structure we're signing up to support long term.

File: core/commands/dht.go (hunk 3 above)
Comment (Stebalien) [reply core/commands/dht.go:171]: Done.

File: core/commands/dht.go (hunk 2 above)
Comment (Stebalien) [reply core/commands/dht.go]: Done.

**Comment by Stebalien** [comment]
1. I've added a comment explaining that this command is unstable.
2. I've moved it to `ipfs stats dht`.

File: core/commands/stat_dht.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, out dhtStat) error {
+			tw := tabwriter.NewWriter(w, 4, 4, 2, ' ', 0)
+			defer tw.Flush()
+
+			// Formats a time into XX ago and remove any decimal
+			// parts. That is, change "2m3.00010101s" to "2m3s ago".
+			now := time.Now()
+			since := func(t time.Time) string {
+				return string(decimalRegex.ReplaceAll(
```
Comment (willscott) [inline core/commands/stat_dht.go:145]: locally, if i do
`now.Sub(t).Round(time.Second).String()` i don't see a decimal. When is this regex needed?

Comment (Stebalien) [reply core/commands/stat_dht.go:145]: You're right... that's a lot simpler.

**Pull request merged**